pandas/_libs/lib.pyx

```diff
@@ -2920,3 +2920,41 @@ def to_object_array_tuples(rows: object) -> np.ndarray:
                 result[i, j] = row[j]
 
     return result
+
+
+def is_bool_list(obj: list) -> bool:
+    """
+    Check if this list contains only bool or np.bool_ objects.
+
+    This is appreciably faster than checking `np.array(obj).dtype == bool`
+
+    obj1 = [True, False] * 100
+    obj2 = obj1 * 100
+    obj3 = obj2 * 100
+    obj4 = [True, None] + obj1
+
+    for obj in [obj1, obj2, obj3, obj4]:
+        %timeit is_bool_list(obj)
+        %timeit np.array(obj).dtype.kind == "b"
+
+    340 ns ± 8.22 ns
+    8.78 µs ± 253 ns
+
+    28.8 µs ± 704 ns
+    813 µs ± 17.8 µs
+
+    3.4 ms ± 168 µs
+    78.4 ms ± 1.05 ms
+
+    48.1 ns ± 1.26 ns
+    8.1 µs ± 198 ns
+    """
+    cdef:
+        object item
+
+    for item in obj:
+        if not util.is_bool_object(item):
+            return False
```

Comment:
Is there any concern with the benchmarks being somewhat misleading due to short circuiting? At least in the Python space similar functions don't seem to benefit much from this

```python
In [6]: %timeit all(isinstance(x, bool) for x in obj)
6.7 µs ± 136 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit np.array(obj).dtype.kind == "b"
5.98 µs ± 121 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
> Is there any concern with the benchmarks being somewhat misleading due to short circuiting?

Only one of the 4 cases posted short-circuits.  Am I misunderstanding the question?

>  At least in the Python space similar functions don't seem to benefit much from this

I was surprised by this too, found that doing it in cython was much more performant after talking with @seberg yesterday about how numpy does this inference.

Reply:
Wow, the difference is big...

NumPy should maybe try to be a bit faster, but that will just make the gap a bit smaller (e.g. by assuming that usually all elements are the same type, I guess a true casting support `np.asarray(..., dtype=bool, casting="equiv")` would fill the gap, but no idea how likely that would be).  Unlike NumPy, you already know what you want here (bools) and what you got (a 1-d list).

Reply:
Ah my bad on short circuiting - misread it. Thanks for comments